modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/toolbarbuttons/plugin.js

```diff
@@ -13,14 +13,246 @@
+
+		if (!table.$.tHead && needRowHeading) {
+			const tableFirstRow = tableBody.getElementsByTag('tr').getItem(0);
+			const tableFirstRowChildCount = tableFirstRow.getChildCount();
+
+			// Change TD to TH:
+
+			for (i = 0; i < tableFirstRowChildCount; i++) {
+				const cell = tableFirstRow.getChild(i);
+
+				// Skip bookmark nodes. (#6155)
+
+				if (
+					cell.type === CKEDITOR.NODE_ELEMENT &&
+					!cell.data('cke-bookmark')
+				) {
+					cell.renameNode('th');
+					cell.setAttribute('scope', 'col');
+				}
+			}
+
+			tableHead = _createElement(table.$.createTHead(), editor);
+			tableHead.append(tableFirstRow.remove());
+		}
+
+		// If we don't need row heading and we have a <thead> element, move the
+		// row out of there and into the <tbody> element.
+
+		if (table.$.tHead !== null && !needRowHeading) {
+
+			// Move the row out of the THead and put it in the TBody:
+
+			tableHead = _createElement(table.$.tHead, editor);
+
+			const previousFirstRow = tableBody.getFirst();
+
+			while (tableHead.getChildCount() > 0) {
+				const newFirstRow = tableHead.getFirst();
+				const newFirstRowChildCount = newFirstRow.getChildCount();
+
+				for (i = 0; i < newFirstRowChildCount; i++) {
+					newCell = newFirstRow.getChild(i);
+
+					if (newCell.type === CKEDITOR.NODE_ELEMENT) {
+						newCell.renameNode('td');
+						newCell.removeAttribute('scope');
+					}
+				}
+
+				newFirstRow.insertBefore(previousFirstRow);
+			}
+
+			tableHead.remove();
+		}
+
+		tableHeading = getHeading(table, editor);
+		const hasColHeading =
+			tableHeading === HEADING_COL || tableHeading === HEADING_BOTH;
+
+		// If we need column heading and the table doesn't have it, convert every first cell in
+		// every row into a `<th scope="row">` element.
+
+		if (!hasColHeading && needColHeading) {
+			for (i = 0; i < table.$.rows.length; i++) {
+				if (table.$.rows[i].cells[0].nodeName.toLowerCase() !== 'th') {
+					newCell = new CKEDITOR.dom.element(
+						table.$.rows[i].cells[0]
+					);
+					newCell.renameNode('th');
+					newCell.setAttribute('scope', 'row');
+				}
+			}
+		}
+
+		// If we don't need column heading but the table has it, convert every first cell in every
+		// row back into a `<td>` element.
+
+		if (hadColHeading && !needColHeading) {
+			for (i = 0; i < table.$.rows.length; i++) {
+				const row = new CKEDITOR.dom.element(table.$.rows[i]);
+
+				if (row.getParent().getName() === 'tbody') {
+					newCell = new CKEDITOR.dom.element(row.$.cells[0]);
+					newCell.renameNode('td');
+					newCell.removeAttribute('scope');
+				}
+			}
+		}
+	}
+
 	CKEDITOR.plugins.add(pluginName, {
 		init(editor) {
+			const headingCommands = [
+				HEADING_NONE,
+				HEADING_ROW,
+				HEADING_COL,
+				HEADING_BOTH,
+			];
+
+			headingCommands.forEach((heading) => {
+				editor.addCommand('tableHeading' + heading, {
+					exec(_editor) {
+						setHeading(null, heading, editor);
+					},
+				});
+			});
```

Comment:
Do we need to expose these commands, so they can be used from outside the plugin? If not, this is probably over-engineering. Instead of `_setHeading` invoking `editor.execCommand`, we can simply use `setHeading(...)` directly.

Reply:
Yes, it makes sense but I decided to left as is considering the issues I was having with other parts of ballooneditor stuff